<a href="https://colab.research.google.com/github/dikshasinghhh/Sentiment-Analysis/blob/main/Sentiment_Analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Natural Language Processing

## Importing the libraries

In [1]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import tensorflow as tf

## Importing the dataset

In [2]:
dataset=pd.read_csv('Restaurant_Reviews.tsv',delimiter='\t', quoting =3)

## Cleaning the texts

In [3]:
import re
import nltk
nltk.download('stopwords')
from nltk.corpus import stopwords
from nltk.stem.porter import PorterStemmer
corpus=[]
for i in range(0,1000):
  review=re.sub('[^a-zA-Z]',' ',dataset['Review'][i])
  review=review.lower()
  review=review.split()
  ps=PorterStemmer()
  all_stopwords= stopwords.words('english')
  all_stopwords.remove('not')
  review=[ps.stem(word) for word in review if not word in set(all_stopwords)]
  review=' '.join(review)
  corpus.append(review)


[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


In [4]:
print (corpus)

['wow love place', 'crust not good', 'not tasti textur nasti', 'stop late may bank holiday rick steve recommend love', 'select menu great price', 'get angri want damn pho', 'honeslti tast fresh', 'potato like rubber could tell made ahead time kept warmer', 'fri great', 'great touch', 'servic prompt', 'would not go back', 'cashier care ever say still end wayyy overpr', 'tri cape cod ravoli chicken cranberri mmmm', 'disgust pretti sure human hair', 'shock sign indic cash', 'highli recommend', 'waitress littl slow servic', 'place not worth time let alon vega', 'not like', 'burritto blah', 'food amaz', 'servic also cute', 'could care less interior beauti', 'perform', 'right red velvet cake ohhh stuff good', 'never brought salad ask', 'hole wall great mexican street taco friendli staff', 'took hour get food tabl restaur food luke warm sever run around like total overwhelm', 'worst salmon sashimi', 'also combo like burger fri beer decent deal', 'like final blow', 'found place accid could not

## Creating the Bag of Words model

In [5]:
from sklearn.feature_extraction.text import CountVectorizer
cv = CountVectorizer(max_features = 1500)
x = cv.fit_transform(corpus).toarray()
y = dataset.iloc[:, -1].values

In [6]:
len(x[0])

1500

## Splitting the dataset into the Training set and Test set

In [7]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=0)

# Initalizing the ANN

# Building the ANN

In [8]:
ann = tf.keras.models.Sequential()

# Adding a hidden layer

In [9]:
ann.add(tf.keras.layers.Dense(units=6,activation='relu'))

# Adding the Second hidden layer

In [10]:
ann.add(tf.keras.layers.Dense(units=6,activation='relu'))

# Adding the output layer

In [11]:
ann.add(tf.keras.layers.Dense(units=1,activation = 'sigmoid'))

# compiling the ANN

In [12]:
ann.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

# Training the ANN on the Training set

In [13]:
ann.fit(x_train,y_train,batch_size = 32, epochs=100)

Epoch 1/100
25/25 [==============================] - 3s 9ms/step - loss: 0.6928 - accuracy: 0.5263
Epoch 2/100
25/25 [==============================] - 0s 11ms/step - loss: 0.6833 - accuracy: 0.6550
Epoch 3/100
25/25 [==============================] - 0s 19ms/step - loss: 0.6671 - accuracy: 0.7500
Epoch 4/100
25/25 [==============================] - 0s 8ms/step - loss: 0.6387 - accuracy: 0.8350
Epoch 5/100
25/25 [==============================] - 0s 9ms/step - loss: 0.5965 - accuracy: 0.8850
Epoch 6/100
25/25 [==============================] - 0s 11ms/step - loss: 0.5401 - accuracy: 0.9038
Epoch 7/100
25/25 [==============================] - 0s 13ms/step - loss: 0.4755 - accuracy: 0.9250
Epoch 8/100
25/25 [==============================] - 0s 8ms/step - loss: 0.4119 - accuracy: 0.9337
Epoch 9/100
25/25 [==============================] - 0s 8ms/step - loss: 0.3529 - accuracy: 0.9425
Epoch 10/100
25/25 [==============================] - 0s 9ms/step - loss: 0.3021 - accuracy: 0.9525
Epoch

# Training the model using different algorithm

## SVM

In [14]:
from sklearn.svm import SVC
classifier = SVC(kernel = 'linear', random_state = 42)
classifier.fit(x_train, y_train)

SVC(kernel='linear', random_state=42)


### Predicting the Test set results

In [15]:
y_pred=classifier.predict(x_test)

### Confusion Matrix

In [16]:
from sklearn.metrics import confusion_matrix,accuracy_score
cm=confusion_matrix(y_test,y_pred)
print(cm)
accuracy_score(y_test,y_pred)

[[79 18]
 [24 79]]


0.79

## 2) NAIVE BAYES

In [17]:
from sklearn.naive_bayes import GaussianNB
classifier = GaussianNB()
classifier.fit(x_train,y_train)

GaussianNB()

### Predicting the test set results

In [18]:
y_pred = classifier.predict(x_test)
print(y_pred)

[1 1 1 0 0 1 1 1 1 1 1 1 1 1 1 1 0 0 0 1 0 0 1 1 1 0 1 1 1 0 1 1 1 1 1 0 1
 0 1 1 1 1 1 0 0 0 1 1 0 0 1 1 1 1 1 0 1 1 0 1 1 0 1 1 1 0 1 1 1 1 1 1 1 1
 0 1 1 0 0 1 0 1 1 0 1 1 1 0 1 1 0 1 0 0 1 1 1 1 1 1 0 1 1 1 0 1 1 1 0 0 0
 1 0 1 1 0 1 1 1 1 1 0 1 1 0 0 1 1 0 1 1 1 0 0 1 1 1 1 1 1 0 1 1 0 1 0 1 1
 1 1 1 0 1 1 1 0 1 1 1 1 1 0 0 1 0 0 1 0 0 0 0 1 1 0 0 1 0 1 0 0 1 0 0 1 0
 1 0 1 0 1 1 0 1 1 1 0 1 1 1 1]


### confusion matrix

In [19]:
from sklearn.metrics import confusion_matrix,accuracy_score
cm=confusion_matrix(y_test,y_pred)
print(cm)
accuracy_score(y_test,y_pred)

[[55 42]
 [12 91]]


0.73

## 3) KNN

In [20]:
from sklearn.neighbors import KNeighborsClassifier
classifier = KNeighborsClassifier(n_neighbors=5,metric='minkowski', p=2)
classifier.fit(x_train,y_train)

KNeighborsClassifier()

###predicting test set results

In [21]:
y_pred=classifier.predict(x_test)

### Confusion Matrix

In [22]:
from sklearn.metrics import confusion_matrix,accuracy_score
cm=confusion_matrix(y_test,y_pred)
print(cm)
accuracy_score(y_test,y_pred)

[[68 29]
 [42 61]]


0.645

## Decision Tree

In [23]:
from sklearn.tree import DecisionTreeClassifier
classifier = DecisionTreeClassifier(criterion='entropy', random_state=0)
classifier.fit(x_train,y_train)

DecisionTreeClassifier(criterion='entropy', random_state=0)

### predicting test set results

In [24]:
y_pred=classifier.predict(x_test)

### Confusion Matrix

In [25]:
from sklearn.metrics import confusion_matrix,accuracy_score
cm=confusion_matrix(y_test,y_pred)
print(cm)
accuracy_score(y_test,y_pred)

[[78 19]
 [31 72]]


0.75

## Random forest

In [26]:
from sklearn.ensemble import RandomForestClassifier
classifier = RandomForestClassifier(n_estimators=10, criterion='entropy', random_state=0)
classifier.fit(x_train,y_train)

RandomForestClassifier(criterion='entropy', n_estimators=10, random_state=0)

### predicting test set results

In [27]:
y_pred=classifier.predict(x_test)

###Confusion Matrix

In [28]:
from sklearn.metrics import confusion_matrix,accuracy_score
cm=confusion_matrix(y_test,y_pred)
print(cm)
accuracy_score(y_test,y_pred)

[[87 10]
 [45 58]]


0.725

# RESULTS: MODEL WITH BEST ACCURACY

```
Model Accuracies:
SVM: 0.79
NAIVE BAYES: 0.73
KNN: 0.64
DECISION TREE: 0.75
RANDOM FOREST: 0.72
```
The best model is SVM with accuracy : 0.79